warehouse/macaroons/auth_policy.py

```diff
@@ -0,0 +1,120 @@
+
+    return auth
+
+
+@implementer(IAuthenticationPolicy)
+class MacaroonAuthenticationPolicy(CallbackAuthenticationPolicy):
+    def __init__(self, callback=None):
+        self.callback = callback
+
+    def unauthenticated_userid(self, request):
+        # If we're calling into this API on a request, then we want to register
+        # a callback which will ensure that the response varies based on the
+        # Authorization header.
+        request.add_response_callback(add_vary_callback("Authorization"))
+
+        # We need to extract our Macaroon from the request.
+        macaroon = extract_http_macaroon(request)
+
+        # Check to see if our Macaroon exists in the database, and if so
+        # fetch the user that is associated with it.
+        macaroon_service = request.find_service(IMacaroonService, context=None)
+        return macaroon_service.find_userid(macaroon)
+
+    def remember(self, request, userid, **kw):
+        # This is a NO-OP because our Macaroon header policy doesn't allow
+        # the ability for authentication to "remember" the user id. This
+        # assumes it has been configured in clients somewhere out of band.
+        return []
+
+    def forget(self, request):
+        # This is a NO-OP because our Macaroon header policy doesn't allow
+        # the ability for authentication to "forget" the user id. This
+        # assumes it has been configured in clients somewhere out of band.
+        return []
+
+
+@implementer(IAuthorizationPolicy)
+class MacaroonAuthorizationPolicy:
+    def __init__(self, policy):
+        self.policy = policy
+
+    def permits(self, context, principals, permission):
+        # The Pyramid API doesn't let us access the request here, so we have to pull it
+        # out of the thread local instead.
+        # TODO: Work with Pyramid devs to figure out if there is a better way to support
+        #       the worklow we are using here or not.
+        request = get_current_request()
+
+        # Our request could possibly be a None, if there isn't an active request, in
+        # that case we're going to always deny, because without a request, we can't
+        # determine if this request is authorized or not.
+        if request is None:
+            return Denied("There was no active request.")
+
+        # Re-extract our Macaroon from the request, it sucks to have to do this work
+        # twice, but I believe it is inevitable unless we pass the Macaroon back as
+        # a principal-- which doesn't seem to be the right fit for it.
+        macaroon = extract_http_macaroon(request)
+
+        # This logic will only happen on requests that are being authenticated with
+        # Macaroons. Any other request will just fall back to the standard Authorization
+        # policy.
+        if macaroon is not None:
+            macaroon_service = request.find_service(IMacaroonService, context=None)
+
+            try:
+                macaroon_service.verify(macaroon)
```

Comment:
I know we talked about how to scope these to just uploads right now, and I wanted to just mention that one way of doing that would be to just one basic caveat (for right now), and basically do this:

```macaroon_service.verify(macaroon, context, principals, permission)```

By passing those three values into the verify function, we could then do something like this to verify:


```python
def verify_caveat(predicate, *, permission):
    # This function has to deal with the entire dictionary that could be possible
    # which means it has to validate the keys we've gotten.
    try:
        data = json.loads(predicate)
    except Exception:  # TODO: Better exception
        return False

    # Check for any unknown predicate keys, we always fail if there are any.
    if data.keys() - {"permissions"}:  # TODO: Might be better to use jsonschema or something eventually.
        return False

    # Go through the caveats and ensure that they match.
    if "permissions" in data and permission not in data["permissions"]:
        # If we have a caveat for permissions, and the permission we're checking for is NOT in that, we fail
        return False

    # Finally, we return True
    return True

v = Verifier()
v.satisfy_general(verify_caveat)
v.verify(...)
```

My suggestion here is basically if we can quickly get some agreement on the general *shape* we want caveats to take (json dict, msgpack dict, whatever) *and* that we want to bake permissions into those caveats, we can quickly  whip up a quick macaaroon based mechanism to handling it.

I think we should still bake version numbers into the ID, just to allow us to completely change the caveat language if we want to, but I think doing this means when we expand, we're already setup to handle the next type of permission we want to add. Upload tokens aren't special, they're just ones that got created with only one permission scope. 

Thoughts?

Reply:
Also in the above code, we could obviously use a class or something instead of a partial, that was just a quick and dirty example of how easy I think it would be to push the verification of "the currently generated tokens are only good for upload" off to the caveat language.

This *would* mean that as implemented an upload token is still not limited to what *projects* it can support and it's scoped the same as the user account that created it in terms of blast radius. That should be pretty easily to do similarly, add a "projects" array to the caveats dictionary and check that the current context is one of those project, basically the same as permissions are done. Everything else should be done by the AuthZ/N framework.

Reply:
> I think we should still bake version numbers into the ID, just to allow us to completely change the caveat language if we want to, but I think doing this means when we expand, we're already setup to handle the next type of permission we want to add. Upload tokens aren't special, they're just ones that got created with only one permission scope.
> 
> Thoughts?

I like this idea! I'll start on it.